Update mermaid 8.5.0 - Refactor repo

tylergibson opened the pull request:
This PR is made in reference to: #39 

This PR addresses or contains:
* Updates to using Mermaid 8.5.0
* Refactors the repo to use new Extension.json schema
* Adds support for all mermaid configuration options
* Enables direct future upstream updates from main Mermaid repo through yarn install

This PR includes:
- [x] Tests (unit/integration)
- [x] CI build passed

Fixes #39

mwjames commented:
Without looking at details:

- Moving files from /src to /includes is a no-go, this repo follows PSR-4
- Moving files outside of /res is not acceptable unless of course
there is a technical reason to do so
- Introducing `yarn` as dependency is questionable, required files
have to be included with a release

On 5/15/20, Tyler Gibson <notifications@github.com> wrote:
> This PR is made in reference to: #39
>
> This PR addresses or contains:
> * Updates to using Mermaid 8.5.0
> * Refactors the repo to use new Extension.json schema
> * Adds support for all mermaid configuration options
> * Enables direct future upstream updates from main Mermaid repo through yarn
> install
>
> This PR includes:
> - [ ] Tests (unit/integration)
> - [ ] CI build passed
>
> Fixes #39
>
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/SemanticMediaWiki/Mermaid/pull/47
>
> -- Commit Summary --
>
>   * First step to 5.8.0 support refactor
>   * Adding custom build of mermaid to fix MW Minifier issues
>   * Adding automated configuration of base mermaid script
>   * Updates all documentation
>
> -- File Changes --
>
>     M .gitignore (14)
>     M DefaultSettings.php (14)
>     D Mermaid.php (194)
>     M README.md (10)
>     M composer.json (15)
>     M docs/INSTALL.md (17)
>     M docs/RELEASE-NOTES.md (11)
>     M docs/USAGE.md (6)
>     M extension.json (53)
>     R i18n/Mermaid.i18n.php (0)
>     A includes/Hooks.php (44)
>     A includes/MermaidParserFunction.php (304)
>     A inject-nomin.js (3)
>     A lib/mermaid.min.js (50)
>     A lib/mermaid.min.js.map (1)
>     R modules/ext.mermaid.js (2)
>     A package.json (56)
>     D res/d3/d3.min.js (8)
>     D res/d3/dagre-d3.min.js (8113)
>     D res/mermaid/7.0.5/mermaid.dark.min.css (1)
>     D res/mermaid/7.0.5/mermaid.forest.min.css (1)
>     D res/mermaid/7.0.5/mermaid.min.css (1)
>     D res/mermaid/7.0.5/mermaid.min.js (28)
>     D res/mermaid/7.0.5/mermaid.neutral.min.css (1)
>     D res/mermaid/7.0.5/mermaidAPI.min.js (1)
>     D res/mermaid/8.0.0/mermaid.dark.min.css (1)
>     D res/mermaid/8.0.0/mermaid.forest.min.css (1)
>     D res/mermaid/8.0.0/mermaid.min.css (1)
>     D res/mermaid/8.0.0/mermaid.min.js (9)
>     D res/mermaid/8.0.0/mermaid.neutral.min.css (1)
>     D res/mermaid/8.0.0/mermaidAPI.min.js (1)
>     D src/HookRegistry.php (116)
>     D src/MermaidParserFunction.php (167)
>     R styles/ext.mermaid.css (0)
>     A yarn.lock (1295)
>
> -- Patch Links --
>
> https://github.com/SemanticMediaWiki/Mermaid/pull/47.patch
> https://github.com/SemanticMediaWiki/Mermaid/pull/47.diff
>
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/SemanticMediaWiki/Mermaid/pull/47
>

tylergibson commented:
Thanks for the feedback @mwjames - I was trying to follow the conventions here - https://www.mediawiki.org/wiki/Manual:Developing_extensions which say to place PHP execution code into /includes. 

Doesn't really matter to me what the dir structure is though :).

Re: Introducing Yarn - The mermaid repo uses Yarn and the mermaid package is an NPM package itself, so you need either npm or yarn (or some other node package manager) to avoid having a static JS dump of Mermaid.js  
This path ensures that, unlike the current release, anyone can reproduce the mermaid.min.js file from origin.

mwjames commented:
> Thanks for the feedback @mwjames - I was trying to follow the conventions
> here - https://www.mediawiki.org/wiki/Manual:Developing_extensions which say
> to place PHP execution code into /includes.
>
> Doesn't really matter to me what the dir structure is though :).

WMF doesn't follow PSR-4 [0] best practice.

> Re: Introducing Yarn - The mermaid repo uses Yarn and the mermaid package is
> an NPM package itself, so you need either npm or yarn (or some other node
> package manager) to avoid having a static JS dump of Mermaid.js
> This path ensures that, unlike the current release, anyone can reproduce the
> mermaid.min.js file from origin.

As development tool `yarn` could be acceptable (e.g. I don't used and
I don't have node.js available) for the release process but we need
the static files which should reside in a subfolder /res (or
/resources) but not scattered around different top directories.

Any new release you may fetch those filesvia yarn and copy them to the
defined location.

[0] https://www.php-fig.org/psr/psr-4/

On 5/15/20, Tyler Gibson <notifications@github.com> wrote:
> Thanks for the feedback @mwjames - I was trying to follow the conventions
> here - https://www.mediawiki.org/wiki/Manual:Developing_extensions which say
> to place PHP execution code into /includes.
>
> Doesn't really matter to me what the dir structure is though :).
>
> Re: Introducing Yarn - The mermaid repo uses Yarn and the mermaid package is
> an NPM package itself, so you need either npm or yarn (or some other node
> package manager) to avoid having a static JS dump of Mermaid.js
> This path ensures that, unlike the current release, anyone can reproduce the
> mermaid.min.js file from origin.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/SemanticMediaWiki/Mermaid/pull/47#issuecomment-629274902

kghbln commented on extension.json:
```diff
@@ -16,10 +17,44 @@
+		"OutputPageParserOutput" : "Mermaid\\Hooks::onOutputPageParserOutput"
+	},
+	"config": {
+		"Theme": "forest"
```
No need for breaking change here. Just use "config_prefix" as `"config_prefix": "mermaidg",` and `"DefaultTheme": "forest"` for "config"

tylergibson commented:
@mwjames I've updated the directory structure to better reflect PSR-4 and PHP-PDP, please have another look.

@kghbln Have made all the changes you noted, have a look and see how they look now

mwjames commented:
As you might be aware this repo comes with PHPUnit tests which needs
to be passed.

Also, if you require a specif config you add a setter/getter because doing
`$globalConfig = MediaWikiServices::getInstance()->getMainConfig();`
defies the idea of dependency injection (you inject whatever you need
from the top).

Doing `$globalConfig =
MediaWikiServices::getInstance()->getMainConfig();` is a kitchen  sink
that cannot be tested so you have to isolate what you need and inject
that instead of just pulling in unnecessary dependencies.

On 5/15/20, Tyler Gibson <notifications@github.com> wrote:
> This PR is made in reference to: #39
>
> This PR addresses or contains:
> * Updates to using Mermaid 8.5.0
> * Refactors the repo to use new Extension.json schema
> * Adds support for all mermaid configuration options
> * Enables direct future upstream updates from main Mermaid repo through yarn
> install
>
> This PR includes:
> - [ ] Tests (unit/integration)
> - [ ] CI build passed
>
> Fixes #39
>
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/SemanticMediaWiki/Mermaid/pull/47
>
> -- Commit Summary --
>
>   * First step to 5.8.0 support refactor
>   * Adding custom build of mermaid to fix MW Minifier issues
>   * Adding automated configuration of base mermaid script
>   * Updates all documentation
>
> -- File Changes --
>
>     M .gitignore (14)
>     M DefaultSettings.php (14)
>     D Mermaid.php (194)
>     M README.md (10)
>     M composer.json (15)
>     M docs/INSTALL.md (17)
>     M docs/RELEASE-NOTES.md (11)
>     M docs/USAGE.md (6)
>     M extension.json (53)
>     R i18n/Mermaid.i18n.php (0)
>     A includes/Hooks.php (44)
>     A includes/MermaidParserFunction.php (304)
>     A inject-nomin.js (3)
>     A lib/mermaid.min.js (50)
>     A lib/mermaid.min.js.map (1)
>     R modules/ext.mermaid.js (2)
>     A package.json (56)
>     D res/d3/d3.min.js (8)
>     D res/d3/dagre-d3.min.js (8113)
>     D res/mermaid/7.0.5/mermaid.dark.min.css (1)
>     D res/mermaid/7.0.5/mermaid.forest.min.css (1)
>     D res/mermaid/7.0.5/mermaid.min.css (1)
>     D res/mermaid/7.0.5/mermaid.min.js (28)
>     D res/mermaid/7.0.5/mermaid.neutral.min.css (1)
>     D res/mermaid/7.0.5/mermaidAPI.min.js (1)
>     D res/mermaid/8.0.0/mermaid.dark.min.css (1)
>     D res/mermaid/8.0.0/mermaid.forest.min.css (1)
>     D res/mermaid/8.0.0/mermaid.min.css (1)
>     D res/mermaid/8.0.0/mermaid.min.js (9)
>     D res/mermaid/8.0.0/mermaid.neutral.min.css (1)
>     D res/mermaid/8.0.0/mermaidAPI.min.js (1)
>     D src/HookRegistry.php (116)
>     D src/MermaidParserFunction.php (167)
>     R styles/ext.mermaid.css (0)
>     A yarn.lock (1295)
>
> -- Patch Links --
>
> https://github.com/SemanticMediaWiki/Mermaid/pull/47.patch
> https://github.com/SemanticMediaWiki/Mermaid/pull/47.diff
>
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/SemanticMediaWiki/Mermaid/pull/47
>

tylergibson commented on extension.json:
```diff
@@ -1,11 +1,12 @@
 	],
 	"url": "https://github.com/SemanticMediaWiki/Mermaid/",
-	"descriptionmsg": "mermaid-desc",
+	"descriptionmsg": "Mediawiki Extension enabling Mermaid.js diagrams at runtime",
```
Thanks for the feedback, should be all sorted now

kghbln commented:
> Have made all the changes you noted, have a look and see how they look now

Looks good to me. Thanks!

mwjames commented:
The way the `elseif` of the config is handled should be isolated as an
extra class to manipulate those available options as parameters.

The new class needs phpunit test.

Overall this should be a 3.0 release given the various changes.

On 5/15/20, Tyler Gibson <notifications@github.com> wrote:
> This PR is made in reference to: #39
>
> This PR addresses or contains:
> * Updates to using Mermaid 8.5.0
> * Refactors the repo to use new Extension.json schema
> * Adds support for all mermaid configuration options
> * Enables direct future upstream updates from main Mermaid repo through yarn
> install
>
> This PR includes:
> - [ ] Tests (unit/integration)
> - [ ] CI build passed
>
> Fixes #39
>
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/SemanticMediaWiki/Mermaid/pull/47
>
> -- Commit Summary --
>
>   * First step to 5.8.0 support refactor
>   * Adding custom build of mermaid to fix MW Minifier issues
>   * Adding automated configuration of base mermaid script
>   * Updates all documentation
>
> -- File Changes --
>
>     M .gitignore (14)
>     M DefaultSettings.php (14)
>     D Mermaid.php (194)
>     M README.md (10)
>     M composer.json (15)
>     M docs/INSTALL.md (17)
>     M docs/RELEASE-NOTES.md (11)
>     M docs/USAGE.md (6)
>     M extension.json (53)
>     R i18n/Mermaid.i18n.php (0)
>     A includes/Hooks.php (44)
>     A includes/MermaidParserFunction.php (304)
>     A inject-nomin.js (3)
>     A lib/mermaid.min.js (50)
>     A lib/mermaid.min.js.map (1)
>     R modules/ext.mermaid.js (2)
>     A package.json (56)
>     D res/d3/d3.min.js (8)
>     D res/d3/dagre-d3.min.js (8113)
>     D res/mermaid/7.0.5/mermaid.dark.min.css (1)
>     D res/mermaid/7.0.5/mermaid.forest.min.css (1)
>     D res/mermaid/7.0.5/mermaid.min.css (1)
>     D res/mermaid/7.0.5/mermaid.min.js (28)
>     D res/mermaid/7.0.5/mermaid.neutral.min.css (1)
>     D res/mermaid/7.0.5/mermaidAPI.min.js (1)
>     D res/mermaid/8.0.0/mermaid.dark.min.css (1)
>     D res/mermaid/8.0.0/mermaid.forest.min.css (1)
>     D res/mermaid/8.0.0/mermaid.min.css (1)
>     D res/mermaid/8.0.0/mermaid.min.js (9)
>     D res/mermaid/8.0.0/mermaid.neutral.min.css (1)
>     D res/mermaid/8.0.0/mermaidAPI.min.js (1)
>     D src/HookRegistry.php (116)
>     D src/MermaidParserFunction.php (167)
>     R styles/ext.mermaid.css (0)
>     A yarn.lock (1295)
>
> -- Patch Links --
>
> https://github.com/SemanticMediaWiki/Mermaid/pull/47.patch
> https://github.com/SemanticMediaWiki/Mermaid/pull/47.diff
>
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/SemanticMediaWiki/Mermaid/pull/47
>

kghbln commented:
> Overall this should be a 3.0 release given the various changes.

I had the same feeling, but this is release details. I will adjust accordingly then doing the next release after the refactor is fully fluff.

kghbln commented:
My approval only covers the comments I made which were resolved in the meantime.

tylergibson commented:
> As you might be aware this repo comes with PHPUnit tests which needs to be passed. Also, if you require a specif config you add a setter/getter because doing `$globalConfig = MediaWikiServices::getInstance()->getMainConfig();` defies the idea of dependency injection (you inject whatever you need from the top). Doing `$globalConfig = MediaWikiServices::getInstance()->getMainConfig();` is a kitchen sink that cannot be tested so you have to isolate what you need and inject that instead of just pulling in unnecessary dependencies.

Ah, I totally forgot about the unit tests.  Sincere apologies.  Also I am extremely rusty with PHP, so any help you may be able to lend with the phpunit setup would be super appreciated.

Re: getMainConfig, this shouldn't be dependency injection at all.  I purposely removed that as there is no need to ever set the mainconfig, only to do a single get-check for the default.  
I am using the MWF defined getter for this, following their prescribed method to do so.
https://www.mediawiki.org/wiki/Manual:Configuration_for_developers
https://www.mediawiki.org/wiki/Manual:Config.php

It looks like we should be using the extension.json to define all of the config features (and to you point, pulling them out of the ifelse insanity I have in there currently)
https://www.mediawiki.org/wiki/Manual:Extension.json/Schema#config

That would get out of the kitchen-sink problem, since it would be a registry for just the scope of the extension, and would provide a known array to iterate over for the values to load into the instance.

I can look at that this weekend unless yall can get to it earlier.

Thanks for all the awesome feedback and help!

mwjames commented:
> Re: getMainConfig, this shouldn't be dependency injection at all. I purposely removed that as there is no need to ever set the mainconfig, only to do a single get-check for the default.
I am using the MWF defined getter for this, following their prescribed method to do so.
https://www.mediawiki.org/wiki/Manual:Configuration_for_developers
https://www.mediawiki.org/wiki/Manual:Config.php

No, `getMainConfig` shouldn't be part of a class (this where you pull in the dependency).

I just looked at the code, all you do is `$globalConfig->has("mermaidgDefaultTheme")` with the config object hence you need to set the value for `mermaidgDefaultTheme` from the outside not from within the class.

Configs are either isolated (by a dedicated class) or you have a setter but never have the entire $GLOBALS or `Config` object be part of a class.

```
		if ( $globalConfig->has("mermaidgDefaultTheme") ) {
			'ext.mermaid.styles'
			$graphConfig = [
		);
				'theme' => $globalConfig->get("mermaidgDefaultTheme")
```

howlowck commented:
@mwjames Please take a look at the updates for this PR.

I made the changes based on your requirements (the best that I could 😬):
* Added `ServiceWiring.php` to include two new services: 
  * `Mermaid.Config` (that gets the default theme)
  * `Mermaid.MermaidConfigExtractor` (that abstracts away the `if/elseif` statements)
* Only Invoked `MediaWikiServices::getInstance()` [in `onParseFunction`](https://github.com/SemanticMediaWiki/Mermaid/pull/47/files#diff-067ac20ecdf4315868de966295e0d01bR54-R57) where we are instantiating the class
* Fixed Tests

Disclamer: I'm a PHP dev, but not a MediaWiki dev, so I'm happy to make any more changes that would make the code more idiomatic to the MediaWiki community.

howlowck commented:
I'm also looking into the failing Travis builds

tylergibson commented:
@kghbln @mwjames - Pinging for review.  Much thanks to my colleague @howlowck - I believe he has addressed the issues and refactored the repo properly.

JeroenDeDauw commented on src/ServiceWiring.php:
```diff
@@ -0,0 +1,32 @@
+	 */
+	'Mermaid.MermaidConfigExtractor' => function (MediaWikiServices $services) {
+		return new MermaidConfigExtractor();
+	},
```
I think using this service wiring stuff is extremely bad practice. Guess you just followed what was recommended on MediaWiki.org though. Lots of stuff to ignore there.

JeroenDeDauw replied:
Tools do not understand what is happening here

howlowck replied:
Hey @JeroenDeDauw, thanks for that.. haha yea. I'm new to mediawiki dev, so I just followed the instructions from mediawiki. Looking for guidance on better practice. We want to change the service binding here, or we can make enhancements once the PR is through.

JeroenDeDauw replied:
A follow up PR is welcome but not required.

JeroenDeDauw commented on tests/phpunit/Unit/MermaidParserFunctionTest.php:
```diff
@@ -13,108 +13,114 @@
 
+	protected function getMockConfig()
+	{
+		$configMock = $parser = $this->createMock('\Mermaid\Config');
```
cp error

JeroenDeDauw replied:
Fixed on master

JeroenDeDauw commented on tests/phpunit/Unit/MermaidParserFunctionTest.php:
```diff
@@ -13,108 +13,114 @@
+	{
+		$valueMap = TestingConsts::EXTRACTOR_VALUE_MAP;
+
+		$extractorMock = $parser = $this->createMock('\Mermaid\MermaidConfigExtractor');
```
cp error

JeroenDeDauw replied:
Fixed on master

JeroenDeDauw commented on tests/bootstrap.php:
```diff
@@ -8,11 +8,11 @@
+//}
 
-print sprintf( "\n%-20s%s\n", "Mermaid: ", $version );
+//print sprintf( "\n%-20s%s\n", "Mermaid: ", $version );
```
Why commented out?

JeroenDeDauw commented on resources/ext.mermaid.js:
```diff
@@ -15,7 +15,7 @@
 
-	mw.loader.using( [ 'mediawiki.api', 'ext.mermaid', 'ext.mermaid.theme.' + config.theme ] ).then( function () {
+	mw.loader.using( [ 'mediawiki.api', 'ext.mermaid' ] ).then( function () {
 
```
What happened with the removed module here?

JeroenDeDauw replied:
Ah I see - it was in one of the removed files.

JeroenDeDauw commented:
Needs a rebase against master due to https://github.com/SemanticMediaWiki/Mermaid/commit/d19485e154580678537648837a6ec6e6c8e06c8f

howlowck commented:
Thanks @JeroenDeDauw! Will rebase today 👍

howlowck commented:
Rebased upstream
